src/platform_impl/linux/wayland/window/shim.rs

```diff
@@ -0,0 +1,388 @@
+    pointers: Vec<WinitPointer>,
+
+    /// Text inputs on the current surface.
+    text_inputs: Vec<TextInputHandler>,
+}
+
+impl WindowHandle {
+    pub fn new(
+        window: Window<ConceptFrame>,
+        size: Arc<Mutex<LogicalSize<u32>>>,
+        pending_window_requests: Arc<Mutex<Vec<WindowRequest>>>,
+    ) -> Self {
+        Self {
+            window,
+            size,
+            pending_window_requests,
+            cursor_icon: Cell::new(CursorIcon::Default),
+            confined: Cell::new(false),
+            cursor_visible: Cell::new(true),
+            pointers: Vec::new(),
+            text_inputs: Vec::new(),
+        }
+    }
+
+    pub fn set_cursor_grab(&self, grab: bool) {
+        // The new requested state matches the current confine status, return.
+        if self.confined.get() == grab {
+            return;
+        }
+
+        self.confined.replace(grab);
+
+        for pointer in self.pointers.iter() {
+            if self.confined.get() {
+                let surface = self.window.surface();
+                pointer.confine(&surface);
+            } else {
+                pointer.unconfine();
+            }
+        }
+    }
+
+    /// Pointer appeared over the window.
+    pub fn pointer_entered(&mut self, pointer: WinitPointer) {
+        let position = self.pointers.iter().position(|p| *p == pointer);
+
+        if position.is_none() {
+            if self.confined.get() {
+                let surface = self.window.surface();
+                pointer.confine(&surface);
+            }
+            self.pointers.push(pointer);
+        }
+
+        // Apply the current cursor style.
+        self.set_cursor_visible(self.cursor_visible.get());
+    }
+
+    /// Pointer left the window.
+    pub fn pointer_left(&mut self, pointer: WinitPointer) {
+        let position = self.pointers.iter().position(|p| *p == pointer);
+
+        if let Some(position) = position {
+            let pointer = self.pointers.remove(position);
+
+            // Drop the confined pointer.
+            if self.confined.get() {
+                pointer.unconfine();
+            }
+        }
+    }
+
+    pub fn text_input_entered(&mut self, text_input: TextInputHandler) {
+        if self
+            .text_inputs
+            .iter()
+            .find(|t| *t == &text_input)
+            .is_none()
+        {
+            self.text_inputs.push(text_input);
+        }
+    }
+
+    pub fn text_input_left(&mut self, text_input: TextInputHandler) {
+        if let Some(position) = self.text_inputs.iter().position(|t| *t == text_input) {
+            self.text_inputs.remove(position);
+        }
+    }
+
+    pub fn set_ime_position(&self, position: LogicalPosition<u32>) {
+        // XXX This won't fly unless user will have a way to request IME window per seat, since
+        // the ime windows will be overlapping, but winit doesn't expose API to specify for
+        // which seat we're setting IME position.
```

Comment:
Should maybe create an issue for that and link to it?

Though it doesn't seem like the end of the world that IME of two seats might sometimes overlap, that sounds quite rare to me?

Reply:
>Though it doesn't seem like the end of the world that IME of two seats might sometimes overlap, that sounds quite rare to me?

The thing is that they **always** overlap, since you set the same position for both windows.

Reply:
I generally dislike meta issues, but it might make sense for seats. Otherwise just an "IME overlaps with multiple seats" should work fine.

The problem's only with multiple seats opening the IME at the same time though, right? Could this lead to an IME opening at the position of the other user's input when an unrelated seat opens it. That would be the bigger problem imho, since multiseat is kinda rare and two seats having IME open at the same time is even rararer.